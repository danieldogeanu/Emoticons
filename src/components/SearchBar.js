import React, { Component } from 'react';
import ClearButton from '../elements/ClearButton';
import '../styles/components/SearchBar.scss';

class SearchBar extends Component {
	constructor(props) {
		super(props);
		this.searchInput = React.createRef();
	}

	selectSearchInput = (e) => {
		if (e.keyCode === 83) {
			this.searchInput.current.focus();
		}
	}

	handleFilterTextChange = (e) => {
		this.props.onFilterTextChange(e.target.value);
		(e.target.value) ? this.ClearBtnCL.add('show') : this.ClearBtnCL.remove('show');
	}

	handleClearInput = (e) => {
		e.preventDefault();
		this.searchInput.current.value = '';
		this.searchInput.current.blur();
		this.props.onFilterTextClear();
		this.ClearBtnCL.remove('show');
	}

	handleKeyUp = (e) => {
		if (e.keyCode === 27) {
			this.handleClearInput(e);
		}
	}

	handleFocus = () => {
		if (!this.SearchInputCL.contains('focused')) this.SearchInputCL.add('focused');
	}

	handleBlur = () => {
		if (this.SearchInputCL.contains('focused')) this.SearchInputCL.remove('focused');
	}

	componentDidMount() {
		window.addEventListener('keyup', this.selectSearchInput);
		this.ClearBtnCL = document.querySelector('.ClearButton').classList;
		this.SearchInputCL = this.searchInput.current.classList;
	}

	componentWillUnmount() {
		window.removeEventListener('keyup', this.selectSearchInput);
	}

	render() {
		return (
			<form className="SearchBar"
				data-testid="SearchBar"
				onSubmit={this.handleClearInput}>
				<input type="text"
					data-testid="SearchInput"
					placeholder="Search Emoticons..."
					value={this.props.filterText}
					onChange={this.handleFilterTextChange}
					onKeyUp={this.handleKeyUp}
					onFocus={this.handleFocus}
					onBlur={this.handleBlur}
					ref={this.searchInput} />
				<ClearButton />
			</form>
		);
	}
}

export default SearchBar;
