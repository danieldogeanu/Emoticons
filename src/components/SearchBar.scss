@import '../styles/includes';

.SearchBar {
	display: block;
	width: 100%;
	margin: 0 0 rem(30) 0;
	padding: 0;
	@include breakpoint(min, rem(301)) { margin: 0 0 rem(35) 0; }
	@include breakpoint(min, rem(481)) { margin: 0 0 rem(40) 0; }
	@include breakpoint(min, rem(851)) { margin: 0 0 rem(50) 0; }

	input[type="text"] {
		appearance: none;
		display: block;
		width: 100%;
		height: rem(40);
		padding: 0 rem(20);
		font-size: rem(15);
		border-radius: rem(6);
		border: rem(1) solid color-scheme(input, border);
		background-color: color-scheme(input, bg);
		color: color-scheme(input, text);
		transition: background-color 0.3s ease, box-shadow 0.3s ease;
		@include box-shadow(0, 0, rem(10), 0, black, 0.1);
		@include breakpoint(min, rem(301)) {
			height: rem(50); font-size: rem(16);
		}
		@include breakpoint(min, rem(481)) {
			height: rem(55); font-size: rem(17);
		}
		@include breakpoint(min, rem(851)) {
			height: rem(60); font-size: rem(18);
		}
		&:focus, &:active {
			background-color: color-scheme(input, bg-focus);
			@include box-shadow(0, 0, rem(20), 0, black, 0.1);
			outline: none;
		}
		@include placeholders() {
			color: color-scheme(input, placeholder);
			opacity: 1;
		}
	}
}