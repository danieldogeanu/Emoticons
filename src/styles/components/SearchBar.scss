@import '../includes';

.SearchBar {
	display: block;
	position: relative;
	width: 100%;
	margin: 0;
	padding: 0;

	input[type="text"] {
		appearance: none;
		display: block;
		width: 100%;
		height: rem(40);
		padding: 0 rem(20);
		font-family: font(open-sans);
		font-weight: weight(regular);
		font-size: rem(15);
		border-radius: rem(6);
		border: 0 none;
		box-shadow: none;
		background-color: color-scheme(input, bg);
		color: color-scheme(input, text);
		transition: background-color 0.3s ease;
		@include breakpoint(min, rem(301)) {
			height: rem(50); font-size: rem(16);
		}
		@include breakpoint(min, rem(481)) {
			height: rem(55); font-size: rem(17);
		}
		@include breakpoint(min, rem(851)) {
			height: rem(60); font-size: rem(18);
		}
		&:focus, &:active {
			background-color: color-scheme(input, bg-focus);
			outline: none;
		}
		@include placeholders() {
			color: color-scheme(input, placeholder);
			opacity: 1;
		}
		&::-ms-clear {
			display: none;
		}
	}
}

.ClearButton {
	display: none;
	position: absolute;
	top: 0; right: 0;
	width: rem(40); height: rem(40);
	background-color: transparent;
	border: 0 none;
	color: color-scheme(input, placeholder);
	transition: color 0.3s ease;
	cursor: pointer;
	&:hover { color: color-scheme(input, text); }
	&:active, &:focus { outline: none; }
	&.show { display: block; }
	.icon.close {
		position: relative; bottom: rem(-2);
		width: rem(20); height: rem(20);
	}
	@include breakpoint(min, rem(301)) {
		width: rem(50); height: rem(50);
		.icon.close { width: rem(22); height: rem(22); }
	}
	@include breakpoint(min, rem(481)) {
		width: rem(55); height: rem(55);
		.icon.close { width: rem(24); height: rem(24); }
	}
	@include breakpoint(min, rem(851)) {
		width: rem(60); height: rem(60);
		.icon.close { width: rem(26); height: rem(26); }
	}
}