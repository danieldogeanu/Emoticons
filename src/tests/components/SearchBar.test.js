import React from 'react';
import {render} from '@testing-library/react';
import SearchBar from '../../components/SearchBar';

const compNames = {
	clear: 'ClearButton',
	search: 'SearchBar',
	input: 'SearchInput',
};

describe('SearchBar Component', () => {

	it('renders search bar properly', () => {
		const {getByTestId, container} = render(<SearchBar />);
		const renderedSearch = getByTestId(compNames.search);
		const renderedInput = getByTestId(compNames.input);
		const renderedClearBtn = container.querySelector(`.${compNames.clear}`);

		expect(renderedSearch).toBeInTheDocument();
		expect(renderedSearch).toHaveClass(compNames.search);
		expect(renderedSearch).toContainElement(renderedInput);
		expect(renderedSearch).toContainElement(renderedClearBtn);
		expect(renderedSearch).toMatchSnapshot();

		expect(renderedInput).toHaveAttribute('placeholder', 'Search Emoticons...');
		expect(renderedInput).toHaveAttribute('type', 'text');
		expect(renderedInput).not.toHaveValue();
	});

});
