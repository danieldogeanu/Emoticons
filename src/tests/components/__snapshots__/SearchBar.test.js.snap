// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar Component renders search bar properly 1`] = `
<form
  class="SearchBar"
  data-testid="SearchBar"
>
  <input
    data-testid="SearchInput"
    placeholder="Search Emoticons..."
    type="text"
    value=""
  />
  <button
    class="ClearButton"
    title="Clear Input"
  >
    <span
      class="show-for-screen-reader"
      data-testid="ScreenReaderText"
    >
      Clear Input
    </span>
    <svg
      class="icon close"
    >
      <use
        xlink:href="#close"
      />
    </svg>
  </button>
</form>
`;
